cdc: support ALTER CHANGEFEED statement for adding/dropping targets

sherman-grewal opened the pull request:
This PR introduces a new SQL statement called ALTER
CHANGEFEED, which allows users to add/drop targets
for an existing changefeed. Note that the changefeed
job must be paused in order for the alterations to
be applied to the changefeed. The syntax of the
statement is as follows:

ALTER CHANGEFEED <job_id> {{ADD|DROP} \<targets\>}...

Where there can be an arbitrary number of ADD/
DROP commands in any order. Once a user
executes this statement, the targets will be
added/dropped, and the statement will return the
job id and the new job description of the changefeed
job. It is also important to note that a user cannot
drop all targets in a changefeed.

Example:
ALTER CHANGEFEED 123 ADD foo,bar DROP baz;

Release note (enterprise change): Added support
for a new SQL statement called ALTER CHANGEFEED,
which allows users to add/drop targets for an
existing changefeed. The syntax of the statement
is as follows:

ALTER CHANGEFEED <job_id> {{ADD|DROP} \<targets\>}...

Where there can be an abritrary number of ADD/DROP
commands in any order. The following is an example
of this statement:

ALTER CHANGEFEED 123 ADD foo,bar DROP baz;

With this statement, users can avoid going through
the process of altering a changefeed on their own,
and rely on CRDB to carry out this task.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/75737)
<!-- Reviewable:end -->

miretskiy reviewed (changes requested):
This is really, really great start.  Few comments; mostly nits.

> Reviewed 11 of 11 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75737#-:-Mv0l1fr7B2cQztch5DN:b-hk3g4f)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @HonoreDB, and @sherman-grewal)

___
*[-- commits, line 6 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75737#-Mv0F2Un71CS81H4tY0k):*
This commit message as well as release note need to be expanded.
Please add some examples; this will help docs writers.
___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 130 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75737#-Mv0jkxF31k6e7G3Qp_P:-Mv0jkxF31k6e7G3Qp_Q:b-wh5e3c) ([raw file](https://github.com/cockroachdb/cockroach/blob/652bc513a9b25c7d467eb465ef59ab646ff34e75/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L130)):*
> ```Go
> 		}
> 
> 		err = saveDetails(ctx, p, jobID, details)
> ```

I think it's very important that the entirety of the logic in this function, *including* the loading of the job, happen under the same transaction provider
in the planner: `p.ExtendedEvalContext().Txn`
If we don't do this, then, it's possible the job is unpaused; or it's possible that the table dropped after we checked its existence, etc.

Loading of the descriptors works w/out txn because we pass in statement time -- which is this txn timestamp, so I think that's fine.
But load/update/save must definitely happen under evalctx.Txn.
___
*[pkg/sql/parser/sql.y, line 4317 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75737#-Mv0FsMh5jy2jxLvMUEZ:-Mv0FsMiEQ5Cob7ASj8L:b-fbv8rw) ([raw file](https://github.com/cockroachdb/cockroach/blob/652bc513a9b25c7d467eb465ef59ab646ff34e75/pkg/sql/parser/sql.y#L4317)):*
> ```Yacc
>     if err != nil {
>       return setErr(sqllex, err)
>     }
> ```

"hardcoding" jobID like this -- vs leaving it for later evaluation makes it impossible to prepare
this statement w/ placeholders.  Perhaps you should just leave it as an expression, and let the hook evaluate it.
___
*[pkg/sql/parser/sql.y, line 4335 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75737#-Mv0GHWv6ohoS-MyURiY:-Mv0GHWv6ohoS-MyURiZ:b-sl11p4) ([raw file](https://github.com/cockroachdb/cockroach/blob/652bc513a9b25c7d467eb465ef59ab646ff34e75/pkg/sql/parser/sql.y#L4335)):*
> ```Yacc
> 
> alter_changefeed_cmd:
>   // ALTER CHANGEFEED <job_id> ADD TARGET ...
> ```

Consider dropping "TARGET".. We don't specify "TARGET"  when we create
changefeed (CREATE CHANGEFEED FOR foo, bar, buz; though we do allow optional TABLE)...
I would probably just use changefeed_targets here?
___
*[pkg/sql/parser/testdata/alter_changefeed, line 5 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75737#-Mv0g4kYCrItESBh2d2N:-Mv0g4kYCrItESBh2d2O:b-csz2zn) ([raw file](https://github.com/cockroachdb/cockroach/blob/652bc513a9b25c7d467eb465ef59ab646ff34e75/pkg/sql/parser/testdata/alter_changefeed#L5)):*
> ```
> ----
> ALTER CHANGEFEED 123 ADD TARGET foo
> ALTER CHANGEFEED 123 ADD TARGET (foo) -- fully parenthesized
> ```

I believe the sytax allows ADD TARGET foo, bar, buz.. Can you check that?
___
*[pkg/sql/parser/testdata/alter_changefeed, line 31 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75737#-Mv0gBRRDaaDHcK7aDno:-Mv0gBRRDaaDHcK7aDnp:bm07ixx) ([raw file](https://github.com/cockroachdb/cockroach/blob/652bc513a9b25c7d467eb465ef59ab646ff34e75/pkg/sql/parser/testdata/alter_changefeed#L31)):*
> ```
> ALTER CHANGEFEED 123 DROP TARGET (foo)  ADD TARGET (bar) -- fully parenthesized
> ALTER CHANGEFEED 123 DROP TARGET foo  ADD TARGET bar -- literals removed
> ALTER CHANGEFEED 123 DROP TARGET _  ADD TARGET _ -- identifiers removed
> ```

Let's also add a mix of add/drops (ALTER... ADD foo DROP bar ADD buz DROP a,b,c)
___
*[pkg/ccl/changefeedccl/alter_changefeed_test.go, line 101 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75737#-Mv0kndy68UDEHcXsMOl:-Mv0kndy68UDEHcXsMOm:b-a50u3r) ([raw file](https://github.com/cockroachdb/cockroach/blob/652bc513a9b25c7d467eb465ef59ab646ff34e75/pkg/ccl/changefeedccl/alter_changefeed_test.go#L101)):*
> ```Go
> 	t.Run(`kafka`, kafkaTest(testFn))
> 	t.Run(`webhook`, webhookTest(testFn))
> 	t.Run(`pubsub`, pubsubTest(testFn))
> ```

I don't think this (or any other test in this file) needs to be sink specific.  You can just execute what you have inside testFn.  If you need to provide sink, provide "null".
___
*[pkg/ccl/changefeedccl/alter_changefeed_test.go, line 134 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75737#-Mv0ken4E4RpuWGr6n4C:-Mv0ken4E4RpuWGr6n4D:b-q8k8a4) ([raw file](https://github.com/cockroachdb/cockroach/blob/652bc513a9b25c7d467eb465ef59ab646ff34e75/pkg/ccl/changefeedccl/alter_changefeed_test.go#L134)):*
> ```Go
> 			fmt.Sprintf(`job %d is not paused`, feed.JobID()),
> 			fmt.Sprintf(`ALTER CHANGEFEED %d ADD TARGET bar`, feed.JobID()),
> 		)
> ```

Would be nice to see what happens if you try to remove all of the targets.


<!-- Sent from Reviewable.io -->

sherman-grewal reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75737)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @HonoreDB, @miretskiy, and @sherman-grewal)

___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 130 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75737#-Mv0jkxF31k6e7G3Qp_P:-MvQKAkQ0XzquQyOTkes:bp1th16) ([raw file](https://github.com/cockroachdb/cockroach/blob/652bc513a9b25c7d467eb465ef59ab646ff34e75/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L130)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I think it's very important that the entirety of the logic in this function, *including* the loading of the job, happen under the same transaction provider
in the planner: `p.ExtendedEvalContext().Txn`
If we don't do this, then, it's possible the job is unpaused; or it's possible that the table dropped after we checked its existence, etc.

Loading of the descriptors works w/out txn because we pass in statement time -- which is this txn timestamp, so I think that's fine.
But load/update/save must definitely happen under evalctx.Txn.
</blockquote></details>

Done, thanks for pointing this out!
___
*[pkg/sql/parser/sql.y, line 4317 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75737#-Mv0FsMh5jy2jxLvMUEZ:-MvQ4bkX0yBgKyF0lvAc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/652bc513a9b25c7d467eb465ef59ab646ff34e75/pkg/sql/parser/sql.y#L4317)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

"hardcoding" jobID like this -- vs leaving it for later evaluation makes it impossible to prepare
this statement w/ placeholders.  Perhaps you should just leave it as an expression, and let the hook evaluate it.
</blockquote></details>

Done.
___
*[pkg/sql/parser/sql.y, line 4335 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75737#-Mv0GHWv6ohoS-MyURiY:-MvL0RsOAdZ9jixBwHl-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/652bc513a9b25c7d467eb465ef59ab646ff34e75/pkg/sql/parser/sql.y#L4335)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Consider dropping "TARGET".. We don't specify "TARGET"  when we create
changefeed (CREATE CHANGEFEED FOR foo, bar, buz; though we do allow optional TABLE)...
I would probably just use changefeed_targets here?
</blockquote></details>

Done.
___
*[pkg/sql/parser/testdata/alter_changefeed, line 5 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75737#-Mv0g4kYCrItESBh2d2N:-MvL8Gou3caY7X9xYZFC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/652bc513a9b25c7d467eb465ef59ab646ff34e75/pkg/sql/parser/testdata/alter_changefeed#L5)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I believe the sytax allows ADD TARGET foo, bar, buz.. Can you check that?
</blockquote></details>

Done.
___
*[pkg/sql/parser/testdata/alter_changefeed, line 31 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75737#-Mv0gBRRDaaDHcK7aDno:-MvL8FESDW_OJXGFJG3F:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/652bc513a9b25c7d467eb465ef59ab646ff34e75/pkg/sql/parser/testdata/alter_changefeed#L31)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Let's also add a mix of add/drops (ALTER... ADD foo DROP bar ADD buz DROP a,b,c)
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/alter_changefeed_test.go, line 101 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75737#-Mv0kndy68UDEHcXsMOl:-MvP3TfjCUeUQK_TaogI:b-t91ljn) ([raw file](https://github.com/cockroachdb/cockroach/blob/652bc513a9b25c7d467eb465ef59ab646ff34e75/pkg/ccl/changefeedccl/alter_changefeed_test.go#L101)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I don't think this (or any other test in this file) needs to be sink specific.  You can just execute what you have inside testFn.  If you need to provide sink, provide "null".
</blockquote></details>

Done. :raised_hands:
___
*[pkg/ccl/changefeedccl/alter_changefeed_test.go, line 134 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75737#-Mv0ken4E4RpuWGr6n4C:-MvP3voi5HEgM5_f8wR7:b-rbrgeq) ([raw file](https://github.com/cockroachdb/cockroach/blob/652bc513a9b25c7d467eb465ef59ab646ff34e75/pkg/ccl/changefeedccl/alter_changefeed_test.go#L134)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Would be nice to see what happens if you try to remove all of the targets.
</blockquote></details>

Yeah it turns out that we cannot remove all of the targets, otherwise it will cause a panic in `StartDistChangefeed`
```
* ERROR: a panic has occurred!
* panic: no spans
* (1) attached stack trace
*   -- stack trace:
*   | runtime.gopanic
*   | 	/usr/local/opt/go/libexec/src/runtime/panic.go:1038
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PartitionSpans
*   | 	/Users/sgrewal/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_physical_planner.go:955
*   | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeeddist.StartDistChangefeed
*   | 	/Users/sgrewal/go/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeeddist/distflow.go:59
*   | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.distChangefeedFlow
*   | 	/Users/sgrewal/go/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeed_dist.go:117
*   | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.(*changefeedResumer).resumeWithRetries
*   | 	/Users/sgrewal/go/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeed_stmt.go:798
*   | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.(*changefeedResumer).Resume
*   | 	/Users/sgrewal/go/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeed_stmt.go:729
*   | github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine.func2
*   | 	/Users/sgrewal/go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1123
*   | github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine
*   | 	/Users/sgrewal/go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1124
*   | github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).runJob
*   | 	/Users/sgrewal/go/src/github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:401
*   | github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).resumeJob.func1
*   | 	/Users/sgrewal/go/src/github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:323
*   | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2
*   | 	/Users/sgrewal/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:494
*   | runtime.goexit
*   | 	/usr/local/opt/go/libexec/src/runtime/asm_amd64.s:1581
* Wraps: (2) panic: no spans
* Error types: (1) *withstack.withStack (2) *errutil.leafError
*
```
For now I will simply error out if someone tries to remove all targets. Thank you for catching this!! :1st_place_medal: 


<!-- Sent from Reviewable.io -->

miretskiy reviewed (approved):
> Reviewed 5 of 8 files at r2, 4 of 4 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75737)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @HonoreDB, and @sherman-grewal)

___
*[-- commits, line 6 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75737#-Mv0F2Un71CS81H4tY0k):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

This commit message as well as release note need to be expanded.
Please add some examples; this will help docs writers.
</blockquote></details>

Just want to make sure you saw this.  Please update PR description with some examples.

This looks great to me.  Giving a <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> .  Perhaps @honoredb might have more comments.
___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 66 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/75737#-MvQZNtlEnoUmGZWWGC4:-MvQZNtlEnoUmGZWWGC5:b-i5njg7) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f3f242fd85392604b0d47ce8d8106c1f128ff79/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L66)):*
> ```Go
> 	if job.Status() != jobs.StatusPaused {
> 		return nil, nil, nil, false, errors.Errorf(`job %d is not paused`, jobID)
> 	}
> ```

I'm not 100% sure, but I think we might want to do everything above (typeExpr, ....) inside fn below -- I think this might allow
job ID specified as a $1 placeholder in the query.  Also... I don't think extendedEvalContext.Txn is set until you actually ... eval -- which is fn.
___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 69 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/75737#-MvQZYPED_zKNiYO1R2K:-MvQZYPED_zKNiYO1R2L:b45di3y) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f3f242fd85392604b0d47ce8d8106c1f128ff79/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L69)):*
> ```Go
> 
> 	header := colinfo.ResultColumns{
> 		{Name: "job_id", Typ: types.Int},
> ```

do we want to return job id, job description (i..e changefeed job after changes)?
___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 177 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/75737#-MvQ_1C7CZJWl0LDNRHC:-MvQ_1C7CZJWl0LDNRHD:bt7bjz4) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f3f242fd85392604b0d47ce8d8106c1f128ff79/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L177)):*
> ```Go
> 		}()
> 
> 		err = p.ExecCfg().JobRegistry.UpdateJobWithTxn(ctx, jobID, p.ExtendedEvalContext().Txn, true, func(
> ```

I don't even know if we should be using read lock.  IN the past "FOR UPDATE" bit us more often than it helped.
So, if you don't specify for update, and you get a conflict (because e.g. somebody else restarted changefeed) -- I think that's fine.
The entire txn retried, and you'll return an error.
So, I would just do something like:
```
const lockForUpdate = false
```
and pass. that to this function.
___
*[pkg/ccl/changefeedccl/alter_changefeed_test.go, line 134 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75737#-Mv0ken4E4RpuWGr6n4C:-MvQYbX4D94GXWW1g0uE:b-j7qk9g) ([raw file](https://github.com/cockroachdb/cockroach/blob/652bc513a9b25c7d467eb465ef59ab646ff34e75/pkg/ccl/changefeedccl/alter_changefeed_test.go#L134)):*
<details><summary><i>Previously, sherman-grewal (Sherman) wrote…</i></summary><blockquote>

Yeah it turns out that we cannot remove all of the targets, otherwise it will cause a panic in `StartDistChangefeed`
```
* ERROR: a panic has occurred!
* panic: no spans
* (1) attached stack trace
*   -- stack trace:
*   | runtime.gopanic
*   | 	/usr/local/opt/go/libexec/src/runtime/panic.go:1038
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PartitionSpans
*   | 	/Users/sgrewal/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_physical_planner.go:955
*   | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeeddist.StartDistChangefeed
*   | 	/Users/sgrewal/go/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeeddist/distflow.go:59
*   | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.distChangefeedFlow
*   | 	/Users/sgrewal/go/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeed_dist.go:117
*   | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.(*changefeedResumer).resumeWithRetries
*   | 	/Users/sgrewal/go/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeed_stmt.go:798
*   | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.(*changefeedResumer).Resume
*   | 	/Users/sgrewal/go/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeed_stmt.go:729
*   | github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine.func2
*   | 	/Users/sgrewal/go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1123
*   | github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine
*   | 	/Users/sgrewal/go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1124
*   | github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).runJob
*   | 	/Users/sgrewal/go/src/github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:401
*   | github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).resumeJob.func1
*   | 	/Users/sgrewal/go/src/github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:323
*   | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2
*   | 	/Users/sgrewal/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:494
*   | runtime.goexit
*   | 	/usr/local/opt/go/libexec/src/runtime/asm_amd64.s:1581
* Wraps: (2) panic: no spans
* Error types: (1) *withstack.withStack (2) *errutil.leafError
*
```
For now I will simply error out if someone tries to remove all spans. Thank you for catching this!! :1st_place_medal: 

</blockquote></details>

Very nice.


<!-- Sent from Reviewable.io -->

HonoreDB reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75737)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @ajwerner, @miretskiy, and @sherman-grewal)

___
*[-- commits, line 6 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75737#-Mv0F2Un71CS81H4tY0k):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Just want to make sure you saw this.  Please update PR description with some examples.

This looks great to me.  Giving a <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> .  Perhaps @honoredb might have more comments.
</blockquote></details>

Thanks for the review! Added a nit but <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> once all comments addressed.
___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 69 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/75737#-MvQZYPED_zKNiYO1R2K:-MvTxtzEDmWU0Zntgqj-:b-9iki1p) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f3f242fd85392604b0d47ce8d8106c1f128ff79/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L69)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

do we want to return job id, job description (i..e changefeed job after changes)?
</blockquote></details>

+1. job_id is arguably redundant with this statement since it can only take a job_id as input right now, but maybe it's worth keeping in anyway for consistency with future versions where it's a name or a where clause/substatement.
___
*[pkg/ccl/changefeedccl/changefeed_stmt.go, line 450 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75737#-MvTyhrU5XzfzltRZJpB:-MvTyhrU5XzfzltRZJpC:b-2i6917) ([raw file](https://github.com/cockroachdb/cockroach/blob/652bc513a9b25c7d467eb465ef59ab646ff34e75/pkg/ccl/changefeedccl/changefeed_stmt.go#L450)):*
> ```Go
> 	}
> 	if !ok {
> 		return pgerror.New(pgcode.InsufficientPrivilege, "permission denied to create changefeed")
> ```

This error string is too specific now, and was never maximally helpful anyway. Suggest changing it to "current user must have a role WITH CONTROLCHANGEFEED".


<!-- Sent from Reviewable.io -->

sherman-grewal reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75737)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @ajwerner, @HonoreDB, @miretskiy, and @sherman-grewal)

___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 66 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/75737#-MvQZNtlEnoUmGZWWGC4:-MvUB8WaBQc-sWYePdpo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f3f242fd85392604b0d47ce8d8106c1f128ff79/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L66)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I'm not 100% sure, but I think we might want to do everything above (typeExpr, ....) inside fn below -- I think this might allow
job ID specified as a $1 placeholder in the query.  Also... I don't think extendedEvalContext.Txn is set until you actually ... eval -- which is fn.
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 69 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/75737#-MvQZYPED_zKNiYO1R2K:-MvU8j4rEkpIcOC5urpx:b-yjtlmo) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f3f242fd85392604b0d47ce8d8106c1f128ff79/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L69)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

+1. job_id is arguably redundant with this statement since it can only take a job_id as input right now, but maybe it's worth keeping in anyway for consistency with future versions where it's a name or a where clause/substatement.
</blockquote></details>

Done. Thanks for catching this! I decided to go with returning the job_id and the jod_description. As @honoredb mentioned, this would likely ensure that we are consistent with future versions where multiple jobs can be edited in one statement
___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 177 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/75737#-MvQ_1C7CZJWl0LDNRHC:-MvU9pB1DeOfnMyDqTNE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f3f242fd85392604b0d47ce8d8106c1f128ff79/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L177)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I don't even know if we should be using read lock.  IN the past "FOR UPDATE" bit us more often than it helped.
So, if you don't specify for update, and you get a conflict (because e.g. somebody else restarted changefeed) -- I think that's fine.
The entire txn retried, and you'll return an error.
So, I would just do something like:
```
const lockForUpdate = false
```
and pass. that to this function.
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/changefeed_stmt.go, line 450 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75737#-MvTyhrU5XzfzltRZJpB:-MvU5Dz903BPCbTyUINt:bt4ox15) ([raw file](https://github.com/cockroachdb/cockroach/blob/652bc513a9b25c7d467eb465ef59ab646ff34e75/pkg/ccl/changefeedccl/changefeed_stmt.go#L450)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

This error string is too specific now, and was never maximally helpful anyway. Suggest changing it to "current user must have a role WITH CONTROLCHANGEFEED".
</blockquote></details>

Done. Thanks for the suggestion!


<!-- Sent from Reviewable.io -->

sherman-grewal commented:
bors r+

sherman-grewal commented:
bors r+

Commit:
changefeedccl: support ALTER CHANGEFEED statement for
adding/dropping targets

References issue #75895

This PR introduces a new SQL statement called ALTER
CHANGEFEED, which allows users to add/drop targets
for an existing changefeed. Note that the changefeed
job must be paused in order for the alterations to
be applied to the changefeed. The syntax of the
statement is as follows:

ALTER CHANGEFEED <job_id> {{ADD|DROP} <targets>}...

Where there can be an arbitrary number of ADD/
DROP commands in any order. Once a user
executes this statement, the targets will be
added/dropped, and the statement will return the
job id and the new job description of the changefeed
job. It is also important to note that a user cannot
drop all targets in a changefeed.

Example:
ALTER CHANGEFEED 123 ADD foo,bar DROP baz;

Release note (enterprise change): Added support
for a new SQL statement called ALTER CHANGEFEED,
which allows users to add/drop targets for an
existing changefeed. The syntax of the statement
is as follows:

ALTER CHANGEFEED <job_id> {{ADD|DROP} <targets>}...

Where there can be an abritrary number of ADD/DROP
commands in any order. The following is an example
of this statement:

ALTER CHANGEFEED 123 ADD foo,bar DROP baz;

With this statement, users can avoid going through
the process of altering a changefeed on their own,
and rely on CRDB to carry out this task.
